### ISSUE TITLE
Terraform S3 backend with AWS SSO profile fails

### ISSUE THREAD

**Issue by lukas-at-harren** [hashicorp/terraform#28872]
### Terraform Version

```
Terraform v0.15.4
on darwin_amd64
```

### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.43.0"
    }
  }

  backend "s3" {
    bucket  = "bucket-redacted"
    key     = "default.tfstate"
    region  = "eu-west-1"
    profile = "profile-redacted"
  }
}
```

### Debug Output

```console
2021-06-03T13:07:28.087+0200 [DEBUG] Adding temp file log sink: /var/folders/t_/fm55d1y11nq3q_mw89b3hfkh0000gp/T/terraform-log961767847
2021-06-03T13:07:28.087+0200 [INFO]  Terraform version: 0.15.4
2021-06-03T13:07:28.087+0200 [INFO]  Go runtime version: go1.16.3
2021-06-03T13:07:28.087+0200 [INFO]  CLI args: []string{"/Users/<redacted>/.bin/terraform", "init"}
2021-06-03T13:07:28.088+0200 [TRACE] Stdout is a terminal of width 104
2021-06-03T13:07:28.088+0200 [TRACE] Stderr is a terminal of width 104
2021-06-03T13:07:28.088+0200 [TRACE] Stdin is a terminal
2021-06-03T13:07:28.088+0200 [DEBUG] Attempting to open CLI config file: /Users/<redacted>/.terraformrc
2021-06-03T13:07:28.088+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2021-06-03T13:07:28.088+0200 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021-06-03T13:07:28.088+0200 [DEBUG] ignoring non-existing provider search directory /Users/<redacted>/.terraform.d/plugins
2021-06-03T13:07:28.088+0200 [DEBUG] ignoring non-existing provider search directory /Users/<redacted>/Library/Application Support/io.terraform/plugins
2021-06-03T13:07:28.088+0200 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2021-06-03T13:07:28.089+0200 [INFO]  CLI command args: []string{"init"}
Initializing modules...
2021-06-03T13:07:28.107+0200 [TRACE] ModuleInstaller: installing child modules for . into .terraform/modules
2021-06-03T13:07:28.110+0200 [DEBUG] Module installer: begin ecs
2021-06-03T13:07:28.115+0200 [TRACE] ModuleInstaller: Module installer: ecs <nil> already installed in ../../modules/aws-ecs-cluster
2021-06-03T13:07:28.115+0200 [DEBUG] Module installer: begin loadbalancer
2021-06-03T13:07:28.120+0200 [TRACE] ModuleInstaller: Module installer: loadbalancer <nil> already installed in ../../modules/aws-alb
2021-06-03T13:07:28.120+0200 [DEBUG] Module installer: begin loadbalancer.cloudflare_security_group
2021-06-03T13:07:28.125+0200 [TRACE] ModuleInstaller: Module installer: loadbalancer.cloudflare_security_group <nil> already installed in ../../modules/aws-cloudflare-security-group
2021-06-03T13:07:28.125+0200 [DEBUG] Module installer: begin networking
2021-06-03T13:07:28.127+0200 [TRACE] ModuleInstaller: Module installer: networking 1.0.0 already installed in .terraform/modules/networking
2021-06-03T13:07:28.127+0200 [DEBUG] Module installer: begin s3_buckets_v1
2021-06-03T13:07:28.129+0200 [TRACE] ModuleInstaller: Module installer: s3_buckets_v1 <nil> already installed in ../../factories/s3-buckets-v1
2021-06-03T13:07:28.129+0200 [DEBUG] Module installer: begin s3_buckets_v1.s3_buckets
2021-06-03T13:07:28.131+0200 [TRACE] ModuleInstaller: Module installer: s3_buckets_v1.s3_buckets <nil> already installed in ../../modules/aws-s3-bucket
2021-06-03T13:07:28.131+0200 [DEBUG] Module installer: begin vpc
2021-06-03T13:07:28.162+0200 [TRACE] ModuleInstaller: Module installer: vpc 3.0.0 already installed in .terraform/modules/vpc
2021-06-03T13:07:28.162+0200 [DEBUG] Module installer: begin web_application_v1
2021-06-03T13:07:28.165+0200 [TRACE] ModuleInstaller: Module installer: web_application_v1 <nil> already installed in ../../factories/web-application-v1
2021-06-03T13:07:28.165+0200 [DEBUG] Module installer: begin web_application_v1.database
2021-06-03T13:07:28.167+0200 [TRACE] ModuleInstaller: Module installer: web_application_v1.database <nil> already installed in ../../modules/aws-rds-postgres
2021-06-03T13:07:28.167+0200 [DEBUG] Module installer: begin web_application_v1.ecs_instances
2021-06-03T13:07:28.170+0200 [TRACE] ModuleInstaller: Module installer: web_application_v1.ecs_instances <nil> already installed in ../../modules/aws-ecs-instance
2021-06-03T13:07:28.170+0200 [DEBUG] Module installer: begin zone
2021-06-03T13:07:28.171+0200 [TRACE] ModuleInstaller: Module installer: zone <nil> already installed in ../../modules/cf-aws-dns-ssl-zone
2021-06-03T13:07:28.171+0200 [DEBUG] Module installer: begin zone.acm
2021-06-03T13:07:28.173+0200 [TRACE] ModuleInstaller: Module installer: zone.acm <nil> already installed in ../../modules/aws-validated-acm-certificate
2021-06-03T13:07:28.173+0200 [TRACE] modsdir: writing modules manifest to .terraform/modules/modules.json

Initializing the backend...
2021-06-03T13:07:28.214+0200 [TRACE] Meta.Backend: built configuration for "s3" backend with hash value 1373428933
2021-06-03T13:07:28.215+0200 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2021-06-03T13:07:28.215+0200 [DEBUG] New state was assigned lineage "f80d6956-89c6-960e-066a-a2955c7c3fc4"
2021-06-03T13:07:28.215+0200 [TRACE] Meta.Backend: moving from default local state only to "s3" backend
2021-06-03T13:07:28.217+0200 [INFO]  Attempting to use session-derived credentials
╷
│ Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.
│
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│
│ Error: NoCredentialProviders: no valid providers in chain. Deprecated.
│       For verbose messaging see aws.Config.CredentialsChainVerboseErrors
│
│
│
╵
```

### Expected Behavior

Terraform should initialise the state on the S3 bucket.

### Actual Behavior

Terraform errored out.

### Steps to Reproduce

* Setup an aws with SSO
* Setup two profiles with SSO
* Use the `provider.tf` from this issue
* Run `TF_LOG=trace terraform init`

### Additional Context

* I am using two AWS SSO profiles.

* The AWS configuration is put in a different folder, than `~/.aws/config` using the "official" environment variable `AWS_CONFIG_FILE`. Works fine for terraform when not using the S3 backend.

* AWS cli works perfectly fine with either of them.

### References

* Similar issue https://github.com/hashicorp/terraform/issues/27650
* @jbardin

**Comment by iJebus** [comment]
If you have that same profile exported as an environment variable, do you still get the error? I feel like I may have bumped into this and that worked, even though it was the same profile.

**Comment by zerthimon** [comment]
Having the same issue with 1.0.0:
```
  backend "s3" {
    profile = "profilename"
  }
```
Gets ignored and uses `[default]` profile instead.

`AWS_PROFILE=profile terraform apply` works correct.

**Comment by exoego** [comment]
I am getting similar error, with different error message.

```
$ terraform -v
Terraform v1.0.0
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.46.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```


```terraform
provider "aws" {
  region = "ap-northeast-1"
  profile = "MY_PROFILE"
  required_version = ">= 0.14.9"
}

terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.44"
    }
  }

  backend "s3" {
    bucket = "MY_BUCKET"
    key    = "terraform"
    region = "ap-northeast-1"
    profile = "MY_PROFILE"
  }
}
```

```
$ terraform init
 terraform plan
╷
│ Error: error configuring S3 Backend: error validating provider credentials: error calling sts:GetCallerIdentity: ExpiredToken: The security token included in the request is expired
│       status code: 403, request id: 4c942602-d5a8-440c-bbf5-554fce92f98b
│ 
│ 
╵
```

However, with `aws2-wrap`, it runs successfully.

```
$ aws2-wrap --profile MY_PROFILE terraform init
...
Terraform has been successfully initialized!
```

**Comment by jaredtbrown** [comment]
I am also experiencing this same issue but on v1.0.5.

**Terraform configuration**
```hcl
terraform {
  backend "s3" {
    bucket  = "BUCKET-REDACTED"
    key     = "KEY-REDACTED"
    region  = "us-east-2"
    encrypt = true
  }

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.27"
    }
  }

  required_version = ">= 0.14.9"
}

provider "aws" {
  profile = "MY-PROFILE"
  region  = "us-east-2"
}
```

The `MY-PROFILE` profile is being ignored and the `[default]` profile is being used instead.

However, just like @iJebus and @zerthimon mentioned, exporting the profile as an environment variable and running initializing the configuration using `AWS_PROFILE=MY-PROFILE terraform init` runs successfully.

**Comment by jaredtbrown** [comment]
Looks like an update with #29017 was released in [1.0.6](https://github.com/hashicorp/terraform/releases/tag/v1.0.6). Perhaps this will fix @lukas-at-harren original issue. 

Upgrading to this version did not fix my problem right away. But I was able to figure out what my issue was.

In looking at the similar issue @lukas-at-harren linked to this one, #27650, I saw someone commented that they did not add the profile to the backend block for their aws configuration. https://github.com/hashicorp/terraform/issues/27650#issuecomment-780251594 I forgot to do the same. Adding the profile to the backend block and upgrading to v1.0.6 fixed my problem.
